### PR TITLE
Add multi-block support

### PR DESCRIPTION
**Pull request by kstribrnAmzn**
### Description of changes:
This change adds support for multiple block
requests and receives. A sequence of up to
NUM_OF_BLOCKS_REQUESTED will be requested
which are sequential in block number if
still needed. Upon receiving the OTA agent
will determine if the block is needed.

With this feature the request data block behavior changes to request up to the first `NUM_OF_BLOCKS_REQUESTED` needed from a starting index. For example...If 10 blocks are needed and block 3 has already been received and the `NUM_OF_BLOCKS_REQUESTED` configured is 3 then the first request would reach out for blocks 0-2 (inclusive). If block 0 and 1 were received then only block 2 would be requested as block 3 is already received. This will not purposely re-request stored blocks.

With this feature upon receiving a new block the block will be stored. Any duplicate blocks are discarded.

`DISCLAIMER`: This is not a very efficient implementation. Often blocks are requested a number of times because we're not tracking what has been requested. This optimization can be added later.

### Issue #, if available:
https://forums.freertos.org/t/bitmap-in-modular-ota-example/19994/4

#### Case 1: NUM_OF_BLOCKS_REQUESTED == 1
```
Successfully connected to IoT Core
OTA Event Queue created.
OTA Event Sent.
OTA Event received 
Received Event is 1 
Request Job Document event Received 
-------------------------------------
OTA Event Sent.
OTA Event received 
Received Event is 2 
Received Job Document event Received 
-------------------------------------
Received OTA Job. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------
Starting The Download. 
Requesting blocks 0-1
OTA Event Sent.
OTA Event received 
Received Event is 5 
Received File Block event Received 
---------------------------------------
Downloaded block 0. Remaining blocks to download: 18. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------
Requesting blocks 1-2
OTA Event Sent.
OTA Event received 
Received Event is 5 
Received File Block event Received 
---------------------------------------
Downloaded block 1. Remaining blocks to download: 17. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------
Requesting blocks 2-3
OTA Event Sent.
OTA Event received 
Received Event is 5 
Received File Block event Received 
---------------------------------------
Downloaded block 2. Remaining blocks to download: 16. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------
Requesting blocks 3-4
OTA Event Sent.
OTA Event received 
Received Event is 5 
Received File Block event Received 
---------------------------------------
Downloaded block 3. Remaining blocks to download: 15. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------
Requesting blocks 4-5
OTA Event Sent.
OTA Event received 
Received Event is 5 
Received File Block event Received 
---------------------------------------
Downloaded block 4. Remaining blocks to download: 14. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------
Requesting blocks 5-6
OTA Event Sent.
OTA Event received 
Received Event is 5 
Received File Block event Received 
---------------------------------------
Downloaded block 5. Remaining blocks to download: 13. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------
Requesting blocks 6-7
OTA Event Sent.
OTA Event received 
Received Event is 5 
Received File Block event Received 
---------------------------------------
...
...
Downloaded block 17. Remaining blocks to download: 1. 
OTA Event Sent.
OTA Event received 
Received Event is 6 
Close file event Received 
-----------------------
Downloaded Data Lorem ipsum dolor sit amet, 
OTA Completed successfully!
```

#### Case 2: NUM_OF_BLOCKS_REQUESTED == 4
```
[WARN] pthread_attr_setstack failed with return value: 22. Default stack will be used.
[WARN] Increase the stack size to PTHREAD_STACK_MIN.
[WARN] pthread_attr_setstack failed with return value: 22. Default stack will be used.
[WARN] Increase the stack size to PTHREAD_STACK_MIN.
[WARN] pthread_attr_setstack failed with return value: 22. Default stack will be used.
[WARN] Increase the stack size to PTHREAD_STACK_MIN.
[WARN] pthread_attr_setstack failed with return value: 22. Default stack will be used.
[WARN] Increase the stack size to PTHREAD_STACK_MIN.
[WARN] pthread_attr_setstack failed with return value: 22. Default stack will be used.
[WARN] Increase the stack size to PTHREAD_STACK_MIN.
Suspend resume task 
D Transport_OpenSSL: Successfully imported root CA.
D Transport_OpenSSL: Successfully imported client certificate.
D Transport_OpenSSL: Successfully imported client certificate private key.
D Transport_OpenSSL: Setting server name a3ixdjl0lkg4lt-ats.iot.us-east-1.amazonaws.com for SNI.
D Transport_OpenSSL: Established a TLS connection.
Successfully connected to IoT Core
OTA Event Queue created.
OTA Event Sent.
OTA Event received 
Received Event is 1 
Request Job Document event Received 
-------------------------------------
OTA Event Sent.
OTA Event received 
Received Event is 2 
Received Job Document event Received 
-------------------------------------
Received OTA Job. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------
Starting The Download. 
Requesting blocks 0-4
OTA Event Sent.
OTA Event received 
Received Event is 5 
Received File Block event Received 
---------------------------------------
Downloaded block 0. Remaining blocks to download: 18. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------
Requesting blocks 1-5
OTA Event Sent.
OTA Event received 
Received Event is 5 
Received File Block event Received 
---------------------------------------
Downloaded block 1. Remaining blocks to download: 17. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------
Requesting blocks 2-6
OTA Event Sent.
OTA Event received 
Received Event is 5 
Received File Block event Received 
---------------------------------------
Downloaded block 2. Remaining blocks to download: 16. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------
Requesting blocks 3-7
OTA Event Sent.
OTA Event received 
Received Event is 5 
Received File Block event Received 
---------------------------------------
Downloaded block 3. Remaining blocks to download: 15. 
OTA Event Sent.
OTA Event received 
Received Event is 4 
Request File Block event Received 
-----------------------------------

...
Downloaded block 17. Remaining blocks to download: 1. 
OTA Event Sent.
OTA Event received 
Received Event is 6 
Close file event Received 
-----------------------
Downloaded Data Lorem ipsum dolor sit amet, 
OTA Completed successfully!
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by kstribrnAmzn** [comment]
Formatting with uncrustify results in significant changes. I'll format at the end once I have approvals to avoid bloating this PR.

File: demo/ota-Agent-Orchestrator/ota_demo.c
Comment (AniruddhaKanhere) [inline demo/ota-Agent-Orchestrator/ota_demo.c:182]: Not a blocker: I think we can coalesce this into one statement which is pretty standard.
```suggestion
    int bitmapSize = ( numOfBlocksRemaining + ( 8 - 1 ) ) / 8;
```

File: demo/ota-Agent-Orchestrator/ota_demo.c
Comment (AniruddhaKanhere) [inline demo/ota-Agent-Orchestrator/ota_demo.c:183]: Maybe memset this to 0?

File: demo/ota-Agent-Orchestrator/ota_demo.c
Comment (AniruddhaKanhere) [inline demo/ota-Agent-Orchestrator/ota_demo.c:318]: NIT: C++ style comments.

File: demo/ota-Agent-Orchestrator/ota_demo.c
Comment (AniruddhaKanhere) [inline demo/ota-Agent-Orchestrator/ota_demo.c:550]: A lot of divisions would happen as this occurs in a loop.
But maybe we can use a simple method to optimize this?

```suggestion
    uint32_t byteIndex = blockId >> 3;
    uint32_t bitIndex = blockId & 0x07;
```

Again, not a blocker as maybe my method is difficult for the reader to understand. :)

File: demo/ota-Agent-Orchestrator/ota_demo.c
Comment (kstribrnAmzn) [reply demo/ota-Agent-Orchestrator/ota_demo.c:182]: I'm going to go with @AniruddhaKanhere's suggestion here. I _really don't like_ those macros in the old OTA agent. It takes something short and simple like the value 8 and aliases it with a macro for readability rather than fixing the core problem, unreadable code.

File: demo/ota-Agent-Orchestrator/ota_demo.c
Comment (kstribrnAmzn) [reply demo/ota-Agent-Orchestrator/ota_demo.c:550]: I don't see this having a big readability impact as the variables in the lines make it clear what is going on. I'll adopt this

File: demo/ota-Agent-Orchestrator/ota_demo.c
Comment (kstribrnAmzn) [reply demo/ota-Agent-Orchestrator/ota_demo.c:318]: Will fix

File: demo/ota-Agent-Orchestrator/ota_demo.c
Comment (kstribrnAmzn) [reply demo/ota-Agent-Orchestrator/ota_demo.c:177]: With https://github.com/aws/aws-iot-core-mqtt-file-streams-embedded-c/commit/690fb2bd10020da916fe54f3e8c59f1e3f925e44 this should now be addressable. I've added the a cfg file to allow for resizing.

**Review by n9wxu — APPROVED** [review]

**Review by AniruddhaKanhere — APPROVED** [review]

**Pull request merged**
